### PR TITLE
Update RHS UI

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
1. Added the status icon to PR's and reviews.
2. Added info about number of approvers on PR.

### CODE REVIEW

File: server/gitlab/api.go [hunk 1]
```diff
@@ -383,6 +465,10 @@ func (g *gitlab) GetUnreads(ctx context.Context, user *UserInfo) ([]*internGitla
 	return notifications, nil
 }
 
+type LabelId struct {
+	ID int `json:"id"`
+}
```
Comment (shivamjosh) [inline server/gitlab/api.go:470]: Where we are using this struct? Remove if unused.

File: webapp/package.json [hunk 2]
```diff
@@ -9,32 +9,52 @@
...
         "prop-types": "15.7.2",
-        "react": "16.8.6",
-        "react-bootstrap": "0.32.4",
-        "react-redux": "5.1.1",
-        "redux": "4.0.1"
+        "react": "16.13.1",
+        "react-bootstrap": "1.3.0",
+        "react-custom-scrollbars": "4.2.1",
+        "react-redux": "7.2.1",
+        "redux": "4.0.5"
```
Comment (shivamjosh) [inline webapp/package.json:21]: Why this change?

File: webapp/package.json [hunk 3]
```diff
@@ -9,32 +9,52 @@
...
+        "css-loader": "4.3.0",
+        "eslint": "7.9.0",
+        "eslint-import-resolver-webpack": "0.12.2",
+        "eslint-plugin-import": "2.22.0",
+        "eslint-plugin-react": "7.20.6",
+        "file-loader": "6.1.0",
+        "style-loader": "1.2.1",
+        "typescript": "4.0.3",
+        "webpack": "4.44.2",
+        "webpack-cli": "3.3.12"
```
Comment (shivamjosh) [inline webapp/package.json:58]: Why this change?

File: webapp/src/components/sidebar_buttons/sidebar_buttons.jsx [hunk 4]
```diff
@@ -3,205 +3,211 @@ import {Tooltip, OverlayTrigger} from 'react-bootstrap';
...
+  };
+
+  openConnectWindow = (e) => {
+      e.preventDefault();
+      window.open(
+          `${this.props.pluginServerRoute}/oauth/connect`,
+          'Connect Mattermost to GitLab',
+          'height=570,width=520',
+      );
+  };
```
Comment (shivamjosh) [inline webapp/src/components/sidebar_buttons/sidebar_buttons.jsx:78]: Revert this change.

File: webapp/src/components/sidebar_buttons/sidebar_buttons.jsx [hunk 5]
```diff
@@ -3,205 +3,211 @@ import {Tooltip, OverlayTrigger} from 'react-bootstrap';
...
+
+      if (!this.props.connected) {
+          if (isTeamSidebar) {
+              return (
+                  <OverlayTrigger
+                      key='gitlabConnectLink'
+                      placement={placement}
+                      overlay={
+                          <Tooltip id='reviewTooltip'>Connect to your GitLab</Tooltip>
+                      }
```
Comment (shivamjosh) [inline webapp/src/components/sidebar_buttons/sidebar_buttons.jsx:106]: Revert. Check everywhere.

File: webapp/src/components/sidebar_right/gitlab_items.tsx [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+import * as CSS from 'csstype';
+import {GitPullRequestIcon, IssueOpenedIcon, IconProps} from '@primer/octicons-react';
+import CrossIcon from "../../images/icons/cross";
+import DotIcon from "../../images/icons/dot";
+import TickIcon from "../../images/icons/tick";
+import SignIcon from '../../images/icons/sign';
+import {formatTimeSince} from '../../utils/date_utils';
+import {GitlabItemsProps, Label, notificationReasons} from "../../types/gitlab_items"
+
+function GitlabItems({ item, theme }: GitlabItemsProps) {
```
Comment (shivamjosh) [inline webapp/src/components/sidebar_right/gitlab_items.tsx:16]:
```suggestion
function GitlabItems({item, theme}: GitlabItemsProps) {
```

File: webapp/src/components/sidebar_right/gitlab_items.tsx [hunk 7]
```diff
@@ -0,0 +1,246 @@
+import React from 'react';
+import {
+  makeStyleFromTheme,
+  changeOpacity,
+} from 'mattermost-redux/utils/theme_utils';
+import { Badge, Tooltip, OverlayTrigger } from "react-bootstrap";
```
Comment (shivamjosh) [inline webapp/src/components/sidebar_right/gitlab_items.tsx:6]:
```suggestion
import {Badge, Tooltip, OverlayTrigger} from "react-bootstrap";
```

File: webapp/src/components/sidebar_right/gitlab_items.tsx [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+          <strong>
+            <a href={item.web_url} target='_blank' rel='noopener noreferrer'>
+              {number}
+            </a>
+          </strong>
+        );
+      }
+    }
+
+    const milestone: JSX.Element | null = item.milestone?(
```
Comment (shivamjosh) [inline webapp/src/components/sidebar_right/gitlab_items.tsx:64]:
```suggestion
    const milestone: JSX.Element | null = item.milestone ? (
```

File: webapp/src/components/sidebar_right/gitlab_items.tsx [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+            ...style.icon,
+            ...((item.created_at || userName) && {
+              paddingLeft: 10,
+            }),
+          }}
+        >
+          <SignIcon />
+          {item.milestone.title}
+        </span>
+      ):null;
```
Comment (shivamjosh) [inline webapp/src/components/sidebar_right/gitlab_items.tsx:77]:
```suggestion
      ) : null;
```

File: webapp/src/components/sidebar_right/gitlab_items.tsx [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+    
+    let labels: JSX.Element[] | null = item.labels?getGitlabLabels(item.labels):null;
```
Comment (shivamjosh) [inline webapp/src/components/sidebar_right/gitlab_items.tsx:79]:
```suggestion
    let labels: JSX.Element[] | null = item.labels ? getGitlabLabels(item.labels) : null;
```

File: webapp/src/components/sidebar_right/gitlab_items.tsx [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+                    default:
+                        status = (
+                            <span
+                                style={{ ...style.icon, ...style.iconFailed }}
+                            >
+                                <CrossIcon />
+                            </span>
+                        );
+                }
+            }
```
Comment (shivamjosh) [inline webapp/src/components/sidebar_right/gitlab_items.tsx:132]: Check indentation.

File: webapp/src/components/sidebar_right/gitlab_items.tsx [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+                        }
+                    >
+                        <i
+                            style={style.conflictIcon}
+                            className="icon icon-alert-outline"
+                        />
+                    </OverlayTrigger>
+                );
+            }
+            let status: JSX.Element | null = null;
```
Comment (shivamjosh) [inline webapp/src/components/sidebar_right/gitlab_items.tsx:102]:
```suggestion

            let status: JSX.Element | null = null;
```

File: webapp/src/components/sidebar_right/gitlab_items.tsx [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+                }
+            }
+        
+      const reviews = (
+        <div style={style.subtitle}>
+            <span className="light">
+                {`${item.approvers} out of ${item.total_reviewers} ${(item.total_reviewers>1?"reviews":"review")} complete.`}
+            </span>
+        </div>
+      )
```
Comment (shivamjosh) [inline webapp/src/components/sidebar_right/gitlab_items.tsx:140]: Indentation.

File: webapp/src/components/sidebar_right/gitlab_items.tsx [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+                                <CrossIcon />
+                            </span>
+                        );
+                }
+            }
+        
+      const reviews = (
+        <div style={style.subtitle}>
+            <span className="light">
+                {`${item.approvers} out of ${item.total_reviewers} ${(item.total_reviewers>1?"reviews":"review")} complete.`}
```
Comment (shivamjosh) [inline webapp/src/components/sidebar_right/gitlab_items.tsx:137]:
```suggestion
                {`${item.approvers} out of ${item.total_reviewers} ${(item.total_reviewers>1 ? "reviews" : "review")} complete.`}
```

File: webapp/src/components/sidebar_right/sidebar_right.tsx [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+            return true;
+        }
+
+        for (let i = 0; i < prs.length; i++) {
+            if (prs[i].id !== prevPrs[i].id) {
+                return true;
+            }
+        }
+    }
+    return false;
```
Comment (shivamjosh) [inline webapp/src/components/sidebar_right/sidebar_right.tsx:77]:
```suggestion
    if (currentState !== targetState) {
        return false;
    }
    
    if (currentState !== prevState) {
        return true;
    }

    if (prs.length !== prevPrs.length) {
        return true;
    }

    for (let i = 0; i < prs.length; i++) {
        if (prs[i].id !== prevPrs[i].id) {
            return true;
        }
    }
```

File: webapp/src/components/sidebar_right/sidebar_right.tsx [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+    reviews: Item[];
+    unreads: Item[],
+    yourPrs: Item[],
+    yourAssignments: Item[],
+    rhsState: string,
+    theme: Theme,
+    actions:any,
+};
+
+export function renderView(props:PropTypes) {
```
Comment (shivamjosh) [inline webapp/src/components/sidebar_right/sidebar_right.tsx:26]: We add a space before defining the type. Change this everywhere.
```suggestion
export function renderView(props: PropTypes) {
```

File: webapp/src/components/sidebar_right/sidebar_right.tsx [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+                              target='_blank'
+                              rel='noopener noreferrer'
+                          >
+                              {title}
+                          </a>
+                      </strong>
+                  </div>
+                  <div>
+                      {!gitlabItems.length?(<div style={style.container}>{'You have no active items'}</div>)
+                      :gitlabItems.map((item)=>
```
Comment (shivamjosh) [inline webapp/src/components/sidebar_right/sidebar_right.tsx:183]:
```suggestion
                      {!gitlabItems.length ?( <div style={style.container}>{'You have no active items'}</div>)
                      : gitlabItems.map((item)=>
```

File: webapp/src/types/gitlab_items.tsx [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+    labels?: Label[];
+    target: Target;
+    approvers:number;
+    total_reviewers:number;
+  }
+  
+  export interface GitlabItemsProps {
+    item: Item;
+    theme: Theme;
+  }
```
Comment (shivamjosh) [inline webapp/src/types/gitlab_items.tsx:76]: EOF missing.

File: webapp/src/types/gitlab_items.tsx [hunk 19]
```diff
@@ -0,0 +1,76 @@
+import {Theme} from 'mattermost-redux/types/preferences';
+import * as CSS from 'csstype';
+
+export const notificationReasons: Record<string, string> = {
```
Comment (shivamjosh) [inline webapp/src/types/gitlab_items.tsx:4]: Fix indentation in the entire file.

File: webapp/src/action_types/index.js [hunk 20]
```diff
@@ -3,9 +3,13 @@ import {id} from '../manifest';
 export default {
     RECEIVED_REVIEWS: `${id}_received_reviews`,
     RECEIVED_YOUR_PRS: `${id}_received_your_prs`,
+    RECEIVED_YOUR_PRS_DETAILS: `${id}_received_your_prs_details`,
```
Comment (avas27JTG) [inline webapp/src/action_types/index.js:6]:
```suggestion
    RECEIVED_YOUR_PR_DETAILS: `${id}_received_your_prs_details`,
```

File: webapp/src/components/sidebar_right/gitlab_items.tsx [hunk 21]
```diff
@@ -0,0 +1,246 @@
+import React from 'react';
+import {
+  makeStyleFromTheme,
+  changeOpacity,
+} from 'mattermost-redux/utils/theme_utils';
+import { Badge, Tooltip, OverlayTrigger } from "react-bootstrap";
+import * as CSS from 'csstype';
+import {GitPullRequestIcon, IssueOpenedIcon, IconProps} from '@primer/octicons-react';
```
Comment (avas27JTG) [inline webapp/src/components/sidebar_right/gitlab_items.tsx:8]: move to top after `react` imports

File: webapp/package.json (hunk 3 above)
Comment (raghavaggarwal2308) [reply webapp/package.json:58]: We matched all the dependency versions with github dependencies.

File: webapp/src/reducers/index.js [hunk 22]
```diff
@@ -115,6 +133,24 @@ function unreads(state = [], action) {
...
+    }
+}
+
+function rhsState(state = null, action) {
+    switch (action.type) {
+    case ActionTypes.UPDATE_RHS_STATE:
+        return action.state;
+    default:
+        return state;
+    }
```
Comment (avas27JTG) [inline webapp/src/reducers/index.js:151]: indentation

File: server/gitlab/api.go [hunk 23]
```diff
@@ -321,37 +345,131 @@ func (g *gitlab) GetYourPrs(ctx context.Context, user *UserInfo) ([]*internGitla
...
+			if res != nil {
+				result = append(result, res)
+			}
+		}(pr.ProjectID, pr.SHA, pr.IID)
+	}
+	wg.Wait()
 	return result, nil
 }
 
+func (g *gitlab) fetchYourPrDetails(c context.Context, log logger.Logger, client *internGitlab.Client, pid int, sha string, iid int) *PRDetails {
```
Comment (shivamjosh) [inline server/gitlab/api.go:403]: Same.

File: server/gitlab/api.go [hunk 24]
```diff
@@ -321,37 +345,131 @@ func (g *gitlab) GetYourPrs(ctx context.Context, user *UserInfo) ([]*internGitla
...
+	var wg sync.WaitGroup
+	for _, pr := range prList {
+		wg.Add(1)
+		go func(pid int, sha string, iid int) {
+			defer wg.Done()
+			res := g.fetchYourPrDetails(ctx, log, client, pid, sha, iid)
+			if res != nil {
+				result = append(result, res)
+			}
+		}(pr.ProjectID, pr.SHA, pr.IID)
```
Comment (shivamjosh) [inline server/gitlab/api.go:397]:
```suggestion
		go func(pid, iid int, sha string) {
			defer wg.Done()
			res := g.fetchYourPrDetails(ctx, log, client, pid, sha, iid)
			if res != nil {
				result = append(result, res)
			}
		}(pr.ProjectID, pr.IID, pr.SHA)
```

File: server/gitlab/api.go [hunk 25]
```diff
@@ -321,37 +345,131 @@ func (g *gitlab) GetYourPrs(ctx context.Context, user *UserInfo) ([]*internGitla
...
+	var approvalDetails *internGitlab.MergeRequestApprovals
+	var err error
+	var resp *internGitlab.Response
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		commitDetails, resp, err = client.Commits.GetCommit(pid, sha, internGitlab.WithContext(c))
+		if respErr := checkResponse(resp); respErr != nil {
+			log.WithError(respErr).Warnf("Failed to fetch commit details for PR")
```
Comment (shivamjosh) [inline server/gitlab/api.go:414]: Add the projectId as well in the error. Change everywhere.

File: webapp/src/reducers/index.js [hunk 26]
```diff
@@ -160,4 +178,6 @@ export default combineReducers({
     gitlabUsers,
     rhsPluginAction,
     rhsState,
+    yourPrsDetails,
```
Comment (avas27JTG) [inline webapp/src/reducers/index.js:181]:
```suggestion
    yourPrDetails,
```

File: webapp/src/reducers/index.js [hunk 27]
```diff
@@ -160,4 +178,6 @@ export default combineReducers({
...
+    reviewsDetails,
```
Comment (avas27JTG) [inline webapp/src/reducers/index.js:182]:
```suggestion
    reviewDetails,
```

why this is repeated/still not fixed? I have already given this comment  that if you use `details` then 1st word must be singular like `reviewDetails`

File: webapp/src/actions/index.js [hunk 28]
```diff
@@ -90,6 +113,28 @@ export function getYourPrs() {
     };
 }
 
+export function getYourPrsDetails(prList) {
```
Comment (avas27JTG) [inline webapp/src/actions/index.js:116]:
```suggestion
export function getYourPrDetails(prList) {
```

fix all such

File: server/command_test.go [hunk 29]
```diff
@@ -190,7 +190,7 @@ func TestListWebhookCommand(t *testing.T) {
 	for _, test := range listWebhookCommandTests {
 		t.Run(test.testName, func(t *testing.T) {
 			p := new(Plugin)
-
+
```
Comment (shivamjosh) [inline server/command_test.go:193]: Revert this change.

File: server/gitlab/api.go [hunk 30]
```diff
@@ -372,6 +383,70 @@ func (g *gitlab) GetLabelDetails(client *internGitlab.Client, pid int, labels in
...
+	var result []*PRDetails
+	var wg sync.WaitGroup
+	for _, pr := range prList {
+		wg.Add(1)
+		go func(pid, iid int, sha string) {
+			defer wg.Done()
+			res, err := g.fetchYourPrDetails(ctx, client, pid, sha, iid)
+			if err != nil {
+				log.WithError(err).Warnf("Failed to fetch PR details")
+				return;
```
Comment (shivamjosh) [inline server/gitlab/api.go:401]: Don't return we have to fetch the details of other PRs as well.

File: server/gitlab/api.go [hunk 31]
```diff
@@ -372,6 +383,70 @@ func (g *gitlab) GetLabelDetails(client *internGitlab.Client, pid int, labels in
...
+			}
+			result = append(result, res)
+		}(pr.ProjectID, pr.IID, pr.SHA)
+	}
+	wg.Wait()
+	return result, nil
+}
+
+func (g *gitlab) fetchYourPrDetails(c context.Context, client *internGitlab.Client, pid int, sha string, iid int) (*PRDetails, error) {
```
Comment (shivamjosh) [inline server/gitlab/api.go:410]:
```suggestion
func (g *gitlab) fetchYourPrDetails(c context.Context, client *internGitlab.Client, pid, iid int, sha string) (*PRDetails, error) {
```

File: server/gitlab/api.go [hunk 32]
```diff
@@ -372,6 +383,70 @@ func (g *gitlab) GetLabelDetails(client *internGitlab.Client, pid int, labels in
...
+		defer wg.Done()
+		commitDetails, resp, err = client.Commits.GetCommit(pid, sha, internGitlab.WithContext(c))
+	}()
+	if respErr := checkResponse(resp); respErr != nil {
+		return nil, errors.Errorf("Failed to fetch commit details for PR with project_id %d", pid)
+
+	}
+	if err != nil {
+		return nil, errors.Errorf("Failed to fetch commit details for PR with project_id %d", pid)
+	}
```
Comment (shivamjosh) [inline server/gitlab/api.go:427]: Why we are not checking for errors inside the go function?

File: server/gitlab/api.go [hunk 33]
```diff
@@ -372,6 +383,70 @@ func (g *gitlab) GetLabelDetails(client *internGitlab.Client, pid int, labels in
...
+	go func() {
+		defer wg.Done()
+		approvalDetails, resp, err = client.MergeRequestApprovals.GetConfiguration(pid, iid, internGitlab.WithContext(c))
+	}()
+	if respErr := checkResponse(resp); respErr != nil {
+		return nil, errors.Errorf("Failed to fetch approval details for PR with project_id %d", pid)
+	}
+	if err != nil {
+		return nil, errors.Errorf("Failed to fetch approval details for PR with project_id %d", pid)
+	}
```
Comment (shivamjosh) [inline server/gitlab/api.go:439]: Same.

**Review by shivamjosh — APPROVED** [review]
